torch/_dynamo/guards.py

```diff
@@ -511,6 +521,106 @@ def combine_scopes(left, right):
 
+    """
+    This is a complex bit of logic. The outline here is brief. For a line by line breakdown, see
+    the code comments below.
+
+    The role of this function is to take the current state of symbolic shape guards, tensor ids in the
+    CURRENT dynamo frame, and tensor names (dynamo's frame agnostic tensor reference mechanism, see TensorCheck and
+    guards.cpp for more info) - and produce executable python expressions for addition to our guarded code components
+    that make their way into check_fn.
+
+    We DO NOT create guards based on ids. The IDs act as a lookup for the following mapping:
+
+    dynamo: tensor_name <> tensor_id
+    shape_env: tensor_id <> shape_expr
+
+    This allows us to then create a tensor_name <> shape_expr association for the current frames guards.
+    """
+
+    def _parse_symbolic_shape_expressions(self, tensor_check_names, tensor_check_ids):
+        # Output
+        finished_expressions = []
+
+        # A mapping of tensor_ids to tensor names
+        id_to_name_map = {}
+
+        # TensorReferences to size. See PySimInt.TensorReference for more info.
+        size_parts = []
+
+        # TensorReferences to size. See PySimInt.TensorReference for more info.
+        stride_parts = []
+
+        # We should not have a shape env, or guards if we are not in config.dynamic shapes
+        # But check it anyway.
+        if not config.dynamic_shapes:
+            return finished_expressions
+
+        # tensor_check_names is the primary tensor association mechanism in dynamo.
+        # All other guards installations are driven off of it, so these ones will too.
+        for name in tensor_check_names:
+            tensor_id = tensor_check_ids[name]
+            id_to_name_map[tensor_id] = name
+
+            if tensor_id in self.shape_env.id_to_expr:
+                # If we made it here, this tensor_id is relevant to dynamo guard installation
+                # AND was found in the shape_env
+                tensor_ref_list = self.shape_env.id_to_expr[tensor_id]
+                for tensor_ref in tensor_ref_list:
+                    val = str(tensor_ref.expr)
+                    if val in ("0", "1"):
+                        continue
+                    if tensor_ref.kind == "size":
+                        size_parts.append(tensor_ref)
+                    if tensor_ref.kind == "stride":
+                        stride_parts.append(tensor_ref)
+
+        # Extract all the guard elements out of guards
+        # The guard format, atm, uses tuple position 0 for the expression
+        # and tuple position 1 for a negation. Eventually, these will be collapsed together.
+        expression_and_evaluation = [
+            (guard[0], guard[1]) for guard in self.shape_env.guards
+        ]
+        for expression, evaluation in expression_and_evaluation:
+            # Expression to string so we can do string replacement on it.
+            # Before you cry, this whole file generates python code,
```

Comment:
I cry. You're using str.replace. What if you are doing a substitution over `s1` in a string that has `s10` in it? Bad things, that's what.

Reply:
Good point, this is a real bug.